src/csharp/Microsoft.Spark/Interop/Ipc/JsonSerDe.cs

```diff
@@ -2,70 +2,139 @@
         {
-            var sortedJObject = new JObject();
-            foreach (JProperty property in jObject.Properties().OrderBy(p => p.Name))
+            Debug.Assert(jObject.ValueKind == JsonValueKind.Object);
+
+            var propertyNames = new List<string>();
+            foreach (JsonProperty prop in jObject.EnumerateObject())
             {
-                if (property.Value is JObject elem)
-                {
-                    sortedJObject.Add(property.Name, elem.SortProperties());
-                }
-                else if (property.Value is JArray arrayElem)
-                {
-                    sortedJObject.Add(property.Name, arrayElem.SortProperties());
-                }
-                else
-                {
-                    sortedJObject.Add(property.Name, property.Value);
-                }
+                propertyNames.Add(prop.Name);
             }
+            propertyNames.Sort();
 
-            return sortedJObject;
-        }
+            writer.WriteStartObject();
+            foreach (string name in propertyNames)
+            {
+                writer.WritePropertyName(name);
+                jObject.GetProperty(name).WriteElementHelper(writer);
+            }
+            writer.WriteEndObject();
+        }  
 
         /// <summary>
         /// Extend method to sort items in a JSON array by keys.
         /// </summary>
-        public static JArray SortProperties(this JArray jArray)
+        private static void SortArrayProperties(this JsonElement jArray, Utf8JsonWriter writer)
         {
-            if (jArray.Count == 0)
+            Debug.Assert(jArray.ValueKind == JsonValueKind.Array);
+
+            writer.WriteStartArray();
+            foreach (JsonElement item in jArray.EnumerateArray())
             {
-                return jArray;
+                item.WriteElementHelper(writer);
             }
+            writer.WriteEndArray();
+        }
 
-            var sortedJArray = new JArray();
-            foreach (JToken item in jArray)
+        /// <summary>
+        /// Helper to appropriately write Json types based on their kind.
+        /// </summary>
+        private static void WriteElementHelper(this JsonElement item, Utf8JsonWriter writer)
+        {
+            Debug.Assert(item.ValueKind != JsonValueKind.Undefined);
+
+            if (item.ValueKind == JsonValueKind.Object)
             {
-                if (item is JObject elem)
-                {
-                    sortedJArray.Add(elem.SortProperties());
-                }
-                else if (item is JArray arrayElem)
-                {
-                    sortedJArray.Add(arrayElem.SortProperties());
-                }
+                item.SortObjectProperties(writer);
+            }
+            else if (item.ValueKind == JsonValueKind.Array)
+            {
+                item.SortArrayProperties(writer);
+            }
+            else
+            {
+                item.WriteTo(writer);
```

Comment:
@suhsteve, @imback82 - can you chime in here?

The **previous** code was ignoring elements within an array. Why was that the case? Is it fine to continue ignoring it or should this be fixed?
```C#
string json = "{\"a\":1, \"c\": [1, 3, 5], \"b\": 3}";
using var doc = JsonDocument.Parse(json);
Console.WriteLine(doc.RootElement.SortProperties().ToString()); // {"a":1,"b":3,"c":[]}
```

Reply:
We can keep the previous behavior and do a follow up PR if needed; since this was only used in parsing datatype json, it wasn't needed.